### DISCUSSION TITLE
Thread Error:OSError: [WinError 10038] 在一个非套接字上尝试了一个操作

### DISCUSSION THREAD

**Discussion by alexchan0412** [FreeOpcUa/opcua-asyncio#848] [Q&A]
Good afternoon,I'm come form china
I wrote the code below:
``from opcua import Client
import time
from threading import Thread
import queue
work = queue.Queue(1000)
client = Client("opc.tcp://127.0.0.1:49320")
def task1(): # 定义任务1
try:
while 1:
client.connect()
T = client.get_node('ns=2;s=test.mac.TA1').get_value()
work.put(T)
time.sleep(0.1)
except Exception as e:
print(e)
finally:
client.disconnect()

def task2():
while 1:
time.sleep(0.5)
print(work.get())

def main(): # 定义main函数
t1 = Thread(target=task1) # 定义线程t1，线程任务为调用task1函数，task1函数的参数是6
t2 = Thread(target=task2) # 定义线程t2，线程任务为调用task2函数，task2函数无参数
t1.start() # 开始运行t1线程
t2.start() # 开始运行t2线程

if name == 'main':
main() # 调用main函数`

Error Meessage:
Exception in thread Thread-1:
Traceback (most recent call last):
File "C:\Users\chenh\AppData\Local\Programs\Python\Python39\lib\site-packages\opcua\client\client.py", line 299, in disconnect
self.close_secure_channel()
File "C:\Users\chenh\AppData\Local\Programs\Python\Python39\lib\site-packages\opcua\client\client.py", line 343, in close_secure_channel
return self.uaclient.close_secure_channel()
File "C:\Users\chenh\AppData\Local\Programs\Python\Python39\lib\site-packages\opcua\client\ua_client.py", line 282, in close_secure_channel
return self._uasocket.close_secure_channel()
File "C:\Users\chenh\AppData\Local\Programs\Python\Python39\lib\site-packages\opcua\client\ua_client.py", line 223, in close_secure_channel
future = self._send_request(request, message_type=ua.MessageType.SecureClose)
File "C:\Users\chenh\AppData\Local\Programs\Python\Python39\lib\site-packages\opcua\client\ua_client.py", line 72, in _send_request
self._socket.write(msg)
File "C:\Users\chenh\AppData\Local\Programs\Python\Python39\lib\site-packages\opcua\common\utils.py", line 118, in write
self.socket.sendall(data)
OSError: [WinError 10038] 在一个非套接字上尝试了一个操作。`

How can i fix it?
Now in china,many engineer want use Python in electric design.
But only some people use FreeOpcUA in CSDN
No one knows haw to use it in internet

**Comment by AndreasHeine** [comment]
again:
> your issue does not really look like a issue with this library rather like a issue in using threading module!

"WinError 10038" is not a OPC UA Error!

what i see in your code is you create connections in a while loop each 100ms a new one... this looks like a DDoS-Script for OPC UA Servers ^^

we provide a lot of examples: https://github.com/FreeOpcUa/opcua-asyncio/tree/master/examples

**Comment by alexchan0412** [comment]
> again:
> 
> > your issue does not really look like a issue with this library rather like a issue in using threading module!
> 
> "WinError 10038" is not a OPC UA Error!
> 
> what i see in your code is you create connections in a while loop each 100ms a new one... this looks like a DDoS-Script for OPC UA Servers ^^
> 
> we provide a lot of examples: https://github.com/FreeOpcUa/opcua-asyncio/tree/master/examples

thanks again, I want use threading to move data at same time.
I also find the example,but annotation is simple

**Comment by alexchan0412** [comment]
> again:
> 
> > your issue does not really look like a issue with this library rather like a issue in using threading module!
> 
> "WinError 10038" is not a OPC UA Error!
> 
> what i see in your code is you create connections in a while loop each 100ms a new one... this looks like a DDoS-Script for OPC UA Servers ^^
> 
> we provide a lot of examples: https://github.com/FreeOpcUa/opcua-asyncio/tree/master/examples

change :time.sleep(1)
same error

**Comment by alexchan0412** [comment]
> again:
> 
> > your issue does not really look like a issue with this library rather like a issue in using threading module!
> 
> "WinError 10038" is not a OPC UA Error!
> 
> what i see in your code is you create connections in a while loop each 100ms a new one... this looks like a DDoS-Script for OPC UA Servers ^^
> 
> we provide a lot of examples: https://github.com/FreeOpcUa/opcua-asyncio/tree/master/examples

new message:
"The security token request type is not valid."(BadRequestTypeInvalid)

**Comment by swamper123** [comment]
I would recommened:

1) Since Pythons-GIL is preventing Threads to be executed in parallel and we don't use multiprocessing or subprocessing: [What is even a Coroutine anyway?](https://www.youtube.com/watch?v=GSiZkP7cI80) is a good video about asyncio (which handles Coroutines) in general.

2) You may want to start with our `server-minimal.py` and `client-minimal.py` example. Observe how the structure is and stuff works. Play with it, perhaps try to get your variable to be read from your server.

3) Take a look on the other examples and adapt it to your purpose.

**Reply by alexchan0412** [reply]
> I would recommened:
> 
> 1. Since Pythons-GIL is preventing Threads to be executed in parallel and we don't use multiprocessing or subprocessing: [What is even a Coroutine anyway?](https://www.youtube.com/watch?v=GSiZkP7cI80) is a good video about asyncio (which handles Coroutines) in general.
> 2. You may want to start with our `server-minimal.py` and `client-minimal.py` example. Observe how the structure is and stuff works. Play with it, perhaps try to get your variable to be read from your server.
> 3. Take a look on the other examples and adapt it to your purpose.

I will look the examples.
Ino the CSDN, a Chinese BBS for Programmers，only has four classes.
https://blog.csdn.net/whahu1989/article/details/120166816

**Comment by alexchan0412** [comment]
Hi，I fixed this problem .when I try to read the data ,it will be connect the sever,EVERYTIME!
so, let "the client.connect()" only run once .
